### PR TITLE
docs: how to use the core api

### PR DESCRIPTION
**Pull request by emmettbutler**
This change adds a big docstring explaining how to write code that uses the Core API.

## Checklist

- [x] Change(s) are motivated and described in the PR description.
- [x] Testing strategy is described if automated tests are not included in the PR.
- [x] Risk is outlined (performance impact, potential for breakage, maintainability, etc).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] [Library release note guidelines](https://ddtrace.readthedocs.io/en/stable/releasenotes.html) are followed. If no release note is required, add label `changelog/no-changelog`.
- [x] Documentation is included (in-code, generated user docs, [public corp docs](https://github.com/DataDog/documentation/)).
- [x] Backport labels are set (if [applicable](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting))

### CODE REVIEW

**Comment by emmettbutler** [comment]
@tabgok @christophe-papazian I'm especially interested in your feedback on this

File: ddtrace/internal/core.py [hunk 1]
```diff
@@ -20,6 +20,102 @@
 except ImportError:
     import ddtrace.vendor.contextvars as contextvars  # type: ignore
 
+"""
```
Comment (brettlangdon) [inline ddtrace/internal/core.py:23]: move to top of file?

This won't get added as `core.__doc__` otherwise.

File: ddtrace/internal/core.py [hunk 2]
```diff
@@ -20,6 +20,102 @@
...
+This file implements the Core API, the abstraction layer between Integrations and Product code.
+The Core API enables two primary use cases: maintaining a tree of ``ExecutionContext`` objects
+and dispatching events.
+
+When using the Core API, keep concerns separate between Products and Integrations. Integrations
```
Comment (brettlangdon) [inline ddtrace/internal/core.py:28]:
```suggestion
Using the Core API, keeps concerns separate between Products and Integrations. Integrations
```

? the sentence reads a tad off to me, trying to figure it out if you meant something else?

File: ddtrace/internal/core.py [hunk 3]
```diff
@@ -20,6 +20,102 @@
...
+should not contain any code that references Products (Tracing, AppSec, Spans, WAF, Debugging, et cetera)
+and Product code should never reference the library being integrated with.
```
Comment (brettlangdon) [inline ddtrace/internal/core.py:30]: > Product code should never reference the library being integrated with

What do you mean by this? Should it be clarified? Like, the Tracing code should never import the `flask` module?

**Review by brettlangdon — COMMENTED** [review]
apologies, I didn't finish reviewing, but wanted to submit initial comments.

File: ddtrace/internal/core.py (hunk 2 above)
Comment (emmettbutler) [reply ddtrace/internal/core.py:28]: It's an imperative, like "when you're using the api, you should keep concerns..."

Comment (brettlangdon) [reply ddtrace/internal/core.py:28]: oh, reading again now makes sense, not sure why my brain wasn't working.

File: ddtrace/internal/core.py [hunk 4]
```diff
@@ -1,3 +1,98 @@
...
+and Product code should never reference the library being integrated with (for example by importing ``flask``).
+
+It's helpful to think of the context tree as a Trace with extra data on each Span. It's similar
+to a tree of Spans in that it represents the parts of the execution state that Datadog products
+care about.
+
+This example shows how ``core.context_with_data`` might be used to create a node in this context tree::
+
+
+    def _patched_request(pin, wrapped, instance, args, kwargs):
```
Comment (emmettbutler) [inline ddtrace/internal/core.py:17]: Remove unbound variables

File: ddtrace/internal/core.py [hunk 5]
```diff
@@ -1,3 +1,98 @@
...
+        with core.context_with_data(
+            "flask._patched_request",
```
Comment (emmettbutler) [inline ddtrace/internal/core.py:19]: make a note about how to choose these names

File: ddtrace/internal/core.py [hunk 6]
```diff
@@ -1,3 +1,106 @@
...
+This example shows how ``core.context_with_data`` might be used to create a node in this context tree::
+
+
+    import flask
+
+
+    def _patched_request(pin, wrapped, args, kwargs):
+        with core.context_with_data(
+            "flask._patched_request",
+            name=".".join(("flask", name)),
```
Comment (brettlangdon) [inline ddtrace/internal/core.py:23]: "name" isn't defined here, should we just use a static name instead?

File: docs/contributing-integrations.rst [hunk 7]
```diff
@@ -1,64 +1,53 @@
...
-Your development process might look like this:
-
-  - Research the library or framework that you're instrumenting. Reading
-    through its docs and code examples will reveal what APIs are meaningful to
-    instrument.
-
-  - Copy the skeleton module provided in ``templates/integration`` and replace
-    ``foo`` with the integration name. The integration name typically matches
-    the library or framework being instrumented::
+There's a skeleton module in ``templates/integration`` that can serve as a helpful starting point
```
Comment (brettlangdon) [inline docs/contributing-integrations.rst:10]: ah, we'll probably want to update this for the new API, but can do that later.

**Review by brettlangdon — DISMISSED** [review]

**Review by brettlangdon — DISMISSED** [review]

**Review by brettlangdon — APPROVED** [review]

**Pull request merged**